cpp/neuralnet/nneval.cpp

```diff
@@ -473,11 +475,16 @@ void NNEvaluator::serve(
       outputBuf.push_back(emptyOutput);
     }
 
-    int symmetry = defaultSymmetry;
-    if(doRandomize)
-      symmetry = rand.nextUInt(NNInputs::NUM_SYMMETRY_COMBINATIONS);
+    for(int row = 0; row<numRows; row++) {
+      if(buf.resultBufs[row]->symmetry == -1) {
```

Comment:
This is hard to follow with the magic -1 number (where is it set, what happens to the value now determined, is there risk of accidentally reusing it?).
When is it not -1 at this point?

Reply:
It's set in the constructor as the default. It is optionally set to specify which symmetry you want. If not set (still -1), then we fall back to using the default or the random as before. Except we're doing this for every row in the array individually now, instead of a single global setting for all rows.

Reply:
I have a comment in the constructor to document but I can add one here also if you like. Or we could make a global define like SYMMETRY_NOTSPECIFIED to compare against instead of -1.

Reply:
if this if statement checking against -1 wasn't there, what would go wrong?

Reply:
It'd probably break. The symmetry has to be between 0 and the max number of symmetries.

Old code was setting it to the default, and then doing randomize, both based on config setting. But this was globally for every eval about to take place in the batch. Now we do the same thing, but on every eval in the batch individually, so we can override some of them if we want to.

The defaults could be added within the constructor for MiscNNInputParams but I was doing it all at once here in the same place the old code used to do it. This also avoids setting it and randomizing it, if it'll just get overwritten later.

Reply:
No, I mean is it ALWAYS -1 here, such that the check is redundant (it should just set it)

Reply:
It gets set at places like https://github.com/roy7/KataGo/pull/1/files#diff-bd67f2c1c385d731233befc2b93c3ad19de564e35748f56424a5eef445e6d829R1139 if you have a specific symmetry you want.

Reply:
And https://github.com/roy7/KataGo/pull/1/files#diff-d20a00509040ee72ce58635f85b6ca9fcabf064b3a7178f808aecbfe5e3a9b45R2012

Reply:
I see, so it kind of belongs to that 'else' branch, but that would lead to some duplication, so it's here via some -1 value that indicates that num symmetries=1.

Reply:
The # of symmetries is set in the config file. This point in the code has no idea how many we're using, that isn't the purpose here, it's just we're about to do a nn eval and need to know which symmetry you want.

Like if you are in gtp and do 'kata-raw-nn 5' it will give you symmetry 5. buf.resultBufs[row]->symmetry would be '5' in that case so the == -1 test skips it.

Reply:
Sander, I've replaced -1 with a clear define instead.